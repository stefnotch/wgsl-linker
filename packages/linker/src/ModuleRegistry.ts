import { ParsedRegistry } from "./ParsedRegistry.js";
import { TextExport, TextModule } from "./ParseModule.js";
import { normalize } from "./PathUtil.js";
import { WgslBundle } from "./WgslBundle.js";

export type CodeGenFn = (
  name: string,
  params: Record<string, string>,
) => string;

export interface GeneratorExport {
  name: string;
  args: string[];
  generate: CodeGenFn;
}

/** a named code generation function */
export interface RegisterGenerator {
  /** export name for this generator */
  name: string;

  /** module namespace for this generator */
  moduleName: string;

  /** function to generate code at runtime */
  generate: CodeGenFn;

  /** arguments to pass when importing from this generator */
  args?: string[];
}

/** a single export from a module */
export type ModuleExport = TextModuleExport | GeneratorModuleExport;

export interface TextModuleExport {
  module: TextModule;
  exp: TextExport;
  kind: "text";
}

export interface GeneratorModule {
  kind: "generator";
  modulePath: string;
  exports: GeneratorExport[];
}

export interface GeneratorModuleExport {
  module: GeneratorModule;
  exp: GeneratorExport;
  kind: "function";
}

export interface RegistryParams {
  /** record of file names and wgsl text for modules */
  wgsl?: Record<string, string>;

  /** record of file names and wgsl text for modules */
  libs?: WgslBundle[];

  /** code generation functions */
  generators?: RegisterGenerator[];
}

const libExp = /\/lib\.w[eg]sl/i;

/**
 * A ModuleRegistry collects exportable code fragments, code generator functions.
 *
 * The ModuleRegistry provides everything required for linkWgsl to process
 * #import statements and generate a complete wgsl shader.
 */
export class ModuleRegistry {
  // map from absolute module path to wgsl/wesl src text
  wgslSrc = new Map<string, string>();
  generators = new Map<string, GeneratorModuleExport>();

  constructor(args?: RegistryParams) {
    if (!args) return;
    const { wgsl = {}, libs = [], generators } = args;

    Object.entries(wgsl).forEach(([fileName, src]) =>
      this.wgslSrc.set(relativeToAbsolute(fileName, "_root"), src),
    );

    libs.forEach(({ name, modules }) => {
      Object.entries(modules).forEach(([fileName, src]) => {
        const absPath = relativeToAbsolute(fileName, name);
        const canonPath =
          libExp.test(absPath) ?
            absPath.slice(0, -"/lib.wgsl".length)
          : absPath;
        this.wgslSrc.set(canonPath, src);
      });
    });

    generators?.map(g => this.registerGenerator(g));
  }

  /**
   * Produce a linked wgsl string with all directives processed
   * (e.g. #import'd functions from other modules are inserted into the resulting string).
   * @param moduleName select the module to use as the root source
   * @param runtimeParams runtime parameters for #import/#export values,
   *  template values, and code generation parameters
   */
  link(moduleName: string, runtimeParams: Record<string, any> = {}): string {
    return this.parsed(runtimeParams).link(moduleName);
  }

  /** Parse the text modules in the registry */
  parsed(runtimeParams: Record<string, any> = {}): ParsedRegistry {
    return new ParsedRegistry(this, runtimeParams);
  }

  /** register a function that generates code on demand */
  registerGenerator(reg: RegisterGenerator): void {
    const exp: GeneratorExport = {
      name: reg.name,
      args: reg.args ?? [],
      generate: reg.generate,
    };
    const module: GeneratorModule = {
      kind: "generator",
      modulePath: reg.moduleName,
      exports: [exp],
    };

    this.generators.set(module.modulePath, { kind: "function", module, exp });
  }
}

export function relativeToAbsolute(
  relativePath: string,
  packageName: string,
): string {
  const normalPath = normalize(relativePath);
  const fullPath = `${packageName}/${normalPath}`;
  return fullPath;
}
