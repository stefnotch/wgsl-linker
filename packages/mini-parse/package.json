{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "prepublishOnly": "run-s build",
    "build": "run-s build:main build:util build:minified",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite build --minify --emptyOutDir=false",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.12.2",
    "@typescript-eslint/parser": "^8.12.2",
    "@vitest/ui": "^2.1.4",
    "berry-pretty": "^0.0.5",
    "eslint": "^9.13.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "typescript": "^5.6.3",
    "vite": "^5.4.10",
    "vite-plugin-dts": "^4.3.0",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.1.4"
  }
}