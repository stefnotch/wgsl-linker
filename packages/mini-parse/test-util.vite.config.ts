/// <reference types="vitest" />
import assert from "node:assert";
import { resolve } from "path";
import { LibraryOptions, defineConfig } from "vite";
import baseViteConfig from "./vite.config.js";

const config = baseViteConfig;
assert(config.build);
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-test-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/test-util/index.ts")];
lib.fileName = "testUtil";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);

// {
//   plugins: [
//     tsconfigPaths(),
//     dts(), // generate
//     // visualizer({ brotliSize: true, gzipSize: true }), // generate stats.html size report
//   ],
//   build: {
//     emptyOutDir: false,
//     lib: {
//       name: "mini-parse-testing",
//       formats: ["es"],
//       entry: [resolve(__dirname, "./src/test-util/index.ts")],
//       fileName: "testUtil"
//     },
//     rollupOptions: {
//       // make sure to externalize deps that shouldn't be bundled
//       // into your library
//       external: ['vitest', 'mini-parse'],
//     },
//     minify: false,
//     sourcemap: true,
//   },
// });
