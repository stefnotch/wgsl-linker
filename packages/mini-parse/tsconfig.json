{
  "extends": ["../../tsconfig.options.json"],
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ESNext", "ESNext.AsyncIterable"],
    "paths": {
      "mini-parse": ["./src"],
      "mini-parse/test-util": ["./src/test-util"]
    }
  },
  "include": ["src/**/*", "*.config.ts"],
  "exclude": ["node_modules"]
}
