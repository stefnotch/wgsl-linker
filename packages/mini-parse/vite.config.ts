/// <reference types="vitest" />
import { resolve } from "path";
import { defineConfig } from "vite";
import dts from "vite-plugin-dts";

export default defineConfig({
  plugins: [dts()],
  test: { setupFiles: "./src/test/TestSetup.ts" },
  build: {
    lib: {
      name: "mini-parse",
      entry: [resolve(__dirname, "src/index.ts")],
    },
    minify: false,
    sourcemap: true,
  },
});
