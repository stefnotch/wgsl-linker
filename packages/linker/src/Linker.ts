import {
  AliasElem,
  FnElem,
  GlobalDirectiveElem,
  StructElem,
  StructMemberElem,
  TypeRefElem,
  VarElem,
} from "./AbstractElems.js";
import { refLog } from "./LinkerLogging.js";
import { RegistryParams } from "./ModuleRegistry.js";
import { ParsedRegistry } from "./ParsedRegistry.js";
import { TextModule } from "./ParseModule.js";
import { SliceReplace, sliceReplace } from "./Slicer.js";
import {
  FoundRef,
  GeneratorRef,
  refName,
  TextRef,
  traverseRefs,
} from "./TraverseRefs.js";
import { replaceWords } from "./Util.js";

type DirectiveRef = {
  kind: "dir";
  expMod: TextModule;
  elem: GlobalDirectiveElem;
};

type LoadableRef = TextRef | GeneratorRef | DirectiveRef;

/** TBD a simple */
export function link(args: RegistryParams) {}

/**
 * Produce a linked wgsl string with all directives processed
 * (e.g. #import'd functions from other modules are inserted into the resulting string).
 *
 * @param runtimeParams runtime parameters for #import/#export values,
 *  template values, and code generation parameters
 */
export function linkWgslModule(
  srcModule: TextModule,
  registry: ParsedRegistry,
  extParams: Record<string, any> = {},
): string {
  const refs = findReferences(srcModule, registry); // all recursively referenced structs and fns

  // mix the merge refs into the import/export refs
  const loadRefs = prepRefsMergeAndLoad(refs);

  // convert global directives into LoadableRefs
  const directiveRefs = globalDirectiveRefs(srcModule);

  // extract export texts, rewriting via rename map and exp/imp args
  const extractRefs = [...loadRefs, ...directiveRefs];
  return extractTexts(extractRefs, extParams);
}

/** Find references to elements like structs and fns to import into the linked result.
 * (note that local functions are not listed unless they are referenced)
 */
export function findReferences(
  srcModule: TextModule,
  registry: ParsedRegistry,
): FoundRef[] {
  // map full export name (with generic params from import) to name for linked result
  const visited = new Map<string, string>();

  // set of linked result names (values of visited map)
  const rootNames = new Set<string>();

  // accumulates all elements to add to the linked result
  const found: FoundRef[] = [];

  traverseRefs(srcModule, registry, refVisit);
  return found;

  /**
   * process one reference found by the reference traversal
   *
   * choose a unique name for the reference so that it can be imported into the
   */
  function refVisit(ref: FoundRef): void {
    const fullName = refFullName(ref);
    let linkName = visited.get(fullName);
    if (!linkName) {
      linkName = uniquifyName(ref.proposedName, rootNames);
      visited.set(fullName, linkName);
      rootNames.add(linkName);
      found.push(ref);
    }

    // always set the rename field to make sure we rewrite calls with module path prefixes
    ref.rename = linkName; // TODO only set if necessary
  }
}

/**
 * Calculate a unique name for a top level element like a struct or fn.
 * @param proposedName
 * @param rootNames
 * @returns the unique name (which may be the proposed name if it's so far unique)
 */
function uniquifyName(
  /** proposed name for this fn in the linked results (e.g. import as name) */
  proposedName: string,
  rootNames: Set<string>,
): string {
  let renamed = proposedName;
  let conflicts = 0;

  // create a unique name
  while (rootNames.has(renamed)) {
    renamed = proposedName + conflicts++;
  }

  return renamed;
}

/**
 * unique id for a future root level element in the form:
 *    moduleName.elemName(impParam1, impParam2, ...)
 * We'll eventually give each unique element a unique fn, struct or variable name
 * in the linked source.
 */
export function refFullName(ref: FoundRef): string {
  const expImpArgs = ref.expInfo?.expImpArgs ?? [];
  const impArgs = expImpArgs.map(([, arg]) => arg);
  const argsStr = "(" + impArgs.join(",") + ")";
  return ref.expMod.modulePath + "." + refName(ref) + argsStr;
}

/**
 * Perpare the refs found in the traverse for loading:
 * . sort through found refs
 *
 * @return the set of refs that will be loaded
 */
function prepRefsMergeAndLoad(refs: FoundRef[]): FoundRef[] {
  const textRefs = refs.filter(r => r.kind === "txt") as TextRef[];
  const generatorRefs = refs.filter(r => r.kind === "gen") as GeneratorRef[];

  return [...generatorRefs, ...textRefs];
}

/** construct DirectiveRefs for from globalDirective elements
 * (so that we can use the standard extract path to copy them to the linked output) */
function globalDirectiveRefs(srcModule: TextModule): DirectiveRef[] {
  const directiveRefs = srcModule.globalDirectives.map(e =>
    toDirectiveRef(e, srcModule),
  );
  return directiveRefs;
}

/** convert a global directive element into a DirectiveRef */
function toDirectiveRef(
  elem: GlobalDirectiveElem,
  expMod: TextModule,
): DirectiveRef {
  return {
    kind: "dir",
    elem,
    expMod,
  };
}

// LATER rename imported vars or aliases
function loadOtherElem(
  ref: TextRef | DirectiveRef,
  extParams: Record<string, string>,
): string {
  const { expMod, elem } = ref;
  const typeRefs = (elem as VarElem | AliasElem).typeRefs ?? [];
  const slicing = typeRefSlices(typeRefs);
  const srcMap = sliceReplace(expMod.src, slicing, elem.start, elem.end);
  // LATER propogate srcMap

  return applyExpImp(srcMap.dest, ref, extParams);
}

function loadGeneratedElem(
  ref: GeneratorRef,
  extParams: Record<string, string>,
): string {
  const genExp = ref.expMod.exports.find(e => e.name === ref.name);
  if (!genExp) {
    refLog(ref, "missing generator", ref.name);
    return "//?";
  }
  const fnName = ref.rename ?? ref.proposedName ?? ref.name;
  const params = refExpImp(ref, extParams);

  const text = genExp?.generate(fnName, params);
  return text;
}

/** load exported text for an import */
function extractTexts(
  refs: LoadableRef[],
  extParams: Record<string, string>,
): string {
  return refs
    .map(r => {
      if (r.kind === "gen") {
        return loadGeneratedElem(r, extParams);
      }
      if (r.kind === "txt") {
        const elemKind = r.elem.kind;
        if (elemKind === "fn") {
          return loadFnText(r.elem, r, extParams);
        }
        if (elemKind === "struct") {
          return loadStruct(r, extParams);
        }
        if (elemKind === "var" || elemKind === "alias") {
          return loadOtherElem(r, extParams);
        }
        console.warn("can't extract. unexpected elem kind:", elemKind, r.elem);
      }
      if (r.kind === "dir") {
        return loadOtherElem(r, extParams);
      }
    })
    .join("\n\n");
}

/** load a struct text, mixing in any elements from #extends */
function loadStruct(ref: TextRef, extParams: Record<string, string>): string {
  const structElem = ref.elem as StructElem;

  const rootMembers =
    structElem.members?.map(m => loadMemberText(m, ref, extParams)) ?? [];

  const allMembers = rootMembers.flat().map(m => "  " + m);
  const membersText = allMembers.join(",\n");
  const name = ref.rename || structElem.name;
  return `struct ${name} {\n${membersText}\n}`;
}

function loadMemberText(
  member: StructMemberElem,
  ref: TextRef,
  extParams: Record<string, string>,
): string {
  const newRef = { ...ref, elem: member };
  return loadOtherElem(newRef, extParams);
}

/** get the export/import param map if appropriate for this ref */
function refExpImp(
  ref: FoundRef,
  extParams: Record<string, string>,
): Record<string, string> {
  const expImp = ref.expInfo?.expImpArgs ?? [];
  const entries = expImp.map(([exp, imp]) => {
    if (imp.startsWith("ext.")) {
      const value = extParams[imp.slice(4)];
      if (value) return [exp, value];

      refLog(ref, "missing ext param", imp, extParams);
    }
    return [exp, imp];
  });
  return Object.fromEntries(entries);
}

function loadFnText(
  elem: FnElem,
  ref: TextRef,
  extParams: Record<string, string>,
): string {
  const { rename } = ref;
  const slicing: SliceReplace[] = [];

  if (rename) {
    const { start, end } = elem.nameElem;
    slicing.push({ start, end, replacement: rename });
  }

  elem.calls.forEach(call => {
    const rename = call?.ref?.rename;
    if (rename) {
      const { start, end } = call;
      slicing.push({ start, end, replacement: rename });
    }
  });

  slicing.push(...typeRefSlices(elem.typeRefs));

  const srcMap = sliceReplace(
    ref.expMod.src,
    slicing,
    elem.start,
    elem.end,
  );

  return applyExpImp(srcMap.dest, ref, extParams);
}

/** rewrite the src text according to module templating and exp/imp params */
function applyExpImp(
  src: string,
  ref: TextRef | DirectiveRef,
  extParams: Record<string, string>,
): string {
  const params = ref.kind === "txt" ? refExpImp(ref, extParams) : {};
  return replaceWords(src, params);
}

function typeRefSlices(typeRefs: TypeRefElem[]): SliceReplace[] {
  const slicing: SliceReplace[] = [];
  typeRefs.forEach(typeRef => {
    const rename = typeRef?.ref?.rename;
    if (rename) {
      const { start, end } = typeRef;
      slicing.push({ start, end, replacement: rename });
    }
  });
  return slicing;
}
