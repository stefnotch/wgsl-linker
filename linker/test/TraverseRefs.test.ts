import { _withBaseLogger } from "@wesl/mini-parse";
import { logCatch } from "@wesl/mini-parse/test-util";
import { expect, test } from "vitest";
import { assertSnapshot } from "@std/testing/snapshot";
import { refFullName } from "../Linker.ts";
import { ModuleRegistry } from "../ModuleRegistry.ts";
import { FoundRef, TextRef, refName, traverseRefs } from "../TraverseRefs.ts";

test("traverse a fn to struct ref", () => {
  const src = `
    import ./file1/AStruct;

    fn main() {
      let a:AStruct; 
    }
  `;
  const module1 = `
    export
    struct AStruct {
      x: u32,
    }
  `;

  const refs = traverseTest(src, module1);
  const exp = refs[1] as TextRef;
  expect(exp.kind).toBe("txt");
  expect(exp.elem.kind).toBe("struct");
  expect(exp.elem.name).toBe("AStruct");
});

test("traverse simple gleam style import", () => {
  const main = `
    import bar/foo;
    fn main() { foo(); }
  `;
  const bar = `
    module bar
    export fn foo() { }
  `;
  const refs = traverseTest(main, bar);
  const exp = refs[1] as TextRef;
  expect(exp.kind).toBe("txt");
  expect(exp.elem.kind).toBe("fn");
  expect(exp.elem.name).toBe("foo");
});

test("traverse nested import with params and support fn", () => {
  const src = `
    import foo(u32) from ./file1
    fn bar() {
      foo(8u);
    }
  `;

  const module1 = `
    import zap from ./file2
  
    export (A)
    fn foo(a: A) { 
      support(a);
      zap();
    }

    fn support() {}
  `;

  const module2 = `
    export 
    fn zap() {}
  `;

  const refs = traverseTest(src, module1, module2);
  const first = refs[1] as TextRef;
  const second = refs[2] as TextRef;
  expect(first.kind).toBe("txt");
  expect(first.expInfo?.expImpArgs).toEqual([["A", "u32"]]);
  expect(second.kind).toBe("txt");
  expect(second.elem.name).toBe("support");
});

test.ignore("traverse importing", () => {
  const src = `
    #import foo(A, B) from ./file1
    fn main() {
      foo(k, l);
    } `;

  const module1 = `
    #export(C, D) importing bar(D) from ./file2
    fn foo(c:C, d:D) { bar(d); } `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const refs = traverseTest(src, module1, module2);

  const importingRef = refs[2] as TextRef;
  expect(importingRef.expInfo?.expImpArgs).toEqual([["X", "B"]]);
});

test.ignore("traverse double importing", () => {
  const src = `
    #import foo(A, B) from ./file1
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing bar(D) from ./file2
    fn foo(c:C, d:D) { bar(d); } `;
  const module2 = `
    #export(X) importing zap(X) from ./file3
    fn bar(x:X) { zap(x); } `;
  const module3 = `
    #export(Y) 
    fn zap(y:Y) { } `;

  const refs = traverseTest(src, module1, module2, module3);

  const expImpArgs = refs.flatMap(r => {
    const er = r as TextRef;
    return er ? [er.expInfo?.expImpArgs] : [];
  });
  expect(expImpArgs[2]).toEqual([["X", "B"]]);
  expect(expImpArgs[3]).toEqual([["Y", "B"]]);
});

test.ignore("traverse importing from a support fn", async (ctx) => {
  const src = `
    #import foo(A, B) from ./file1
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing support(D) from ./file1
    fn foo(c:C, d:D) { support(d); } 
    
    #export(D) importing bar(D) from ./file2
    fn support(d:D) { bar(d); }
    `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const refs = traverseTest(src, module1, module2);

  const expImpArgs = refs.flatMap(r => {
    const er = r as TextRef;
    return er ? [{ name: er.elem.name, args: er.expInfo?.expImpArgs }] : [];
  });
  await assertSnapshot(ctx, expImpArgs);
});

test.ignore("traverse importing from a local call fails", () => {
  const src = `
    #import foo(A, B) from ./file1
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing bar(D) 
    fn foo(c:C, d:D) { support(d); } 
    
    fn support(d:D) { bar(d); } //  need to mark this as an export with importing, so we can map params
    `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const { log } = traverseWithLog(src, module1, module2);
  expect(log.length).not.toBe(0);
});

test.ignore("importing args don't match", async (ctx) => {
  const src = `
    #import foo(A, B) from ./file1
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing bar(E) from ./file2
    fn foo(c:C, d:D) { bar(d); } `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const { log } = traverseWithLog(src, module1, module2);

  await assertSnapshot(ctx, log);
});

test("mismatched import export params", async (ctx) => {
  const src = `
    #import foo(A, B) from ./file1
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C) 
    fn foo(c:C) { } `;

  const { log } = traverseWithLog(src, module1);
  await assertSnapshot(ctx, log);
});

test("traverse var to gleam style struct ref", () => {
  const main = `
     import foo/Bar;
     var x: Bar;
     fn main() { }
   `;
  const foo = `
      module foo
      export struct Bar { f: f32 }
   `;

  const refs = traverseTest(main, foo);
  const structRef = refs.find(
    ref => ref.kind === "txt" && ref.elem.kind === "struct",
  );
  expect(structRef).toBeDefined();
});

test("traverse a struct to struct ref", () => {
  const src = `
    #import AStruct from ./file1

    struct SrcStruct {
      a: AStruct,
    }
  `;
  const module1 = `
    #export
    struct AStruct {
      x: u32,
    }
  `;

  const refs = traverseTest(src, module1);
  expect(refs[1].kind).toBe("txt");
  expect(refName(refs[1])).toBe("AStruct");
});

test("traverse a global var to struct ref", () => {
  const src = `
    #import Uniforms from ./file1

    @group(0) @binding(0) var<uniform> u: Uniforms;      
    `;
  const module1 = `
    #export
    struct Uniforms {
      model: mat4x4<f32>,
    }
  `;

  const refs = traverseTest(src, module1);
  const exp = refs[1] as TextRef;
  expect(exp.kind).toBe("txt");
  expect(exp.elem.kind).toBe("struct");
  expect(exp.elem.name).toBe("Uniforms");
});

test("traverse transitive struct refs", () => {
  const src = `
    #import AStruct  from ./file1

    struct SrcStruct {
      a: AStruct,
    }
  `;
  const module1 = `
    #import BStruct from ./file2

    #export
    struct AStruct {
      s: BStruct,
    }
  `;
  const module2 = `
    #export
    struct BStruct {
      x: u32,
    }
  `;

  const refs = traverseTest(src, module1, module2);
  expect(refName(refs[1])).toBe("AStruct");
  expect(refName(refs[2])).toBe("BStruct");
});

test.ignore("traverse #export importing struct to struct", () => {
  const src = `
    #import AStruct(MyStruct)

    struct MyStuct {
      x: u32
    }

    struct HomeStruct {
      a:AStruct
    }
  `;
  const module1 = `
    #export(B) importing BStruct(B)
    struct AStruct {
      b: BStruct
    }
  `;

  const module2 = `
    #export(Y) 
    struct BStruct {
      Y: Y
    }
  `;
  const refs = traverseTest(src, module1, module2);
  expect(refName(refs[2])).toBe("AStruct");
  expect(refName(refs[3])).toBe("BStruct");
});

test("traverse ref from struct constructor", () => {
  const src = `
    #import AStruct from ./file1

    fn main() {
      var x = AStruct(1u);
    }
  `;
  const module1 = `
    #export
    struct AStruct {
      b: u32
    }
  `;

  const refs = traverseTest(src, module1);
  expect(refName(refs[1])).toBe("AStruct");
});

test.ignore("traverse #extends", () => {
  const src = `
    #extends A 
    struct B {
      x: u32
    }
  `;
  const module1 = `
    #export
    struct A {
      z: u32
    }
  `;
  const refs = traverseTest(src, module1);
  expect(refName(refs[1])).toBe("A");
});

test("traverse with local support struct", () => {
  const src = `
    #import A from ./file1

    fn b() { var a: A; var b: B; }

    struct B { x: u32 }
  `;
  const module1 = `
    #export
    struct A { y: i32 }
  `;

  const refs = traverseTest(src, module1);
  const refNames = refs.map(refName);
  expect(refNames).toEqual(["B", "b", "A"]);
});

test("traverse from return type of function", () => {
  const src = `
    #import A from ./file1

    fn b() -> A { }
  `;
  const module1 = `
    #export
    struct A { y: i32 }
  `;

  const refs = traverseTest(src, module1);
  const refNames = refs.map(refName);
  expect(refNames).toEqual(["b", "A"]);
});

test("traverse skips built in fn and type", () => {
  const src = `
    fn foo() {
      bar();
      min(3,4);
      vec3(u);
    }
    fn bar() {}
  `;

  const { refs, log } = traverseWithLog(src);
  const refNames = refs.map(refName);
  // refs.map(r => refLog(r));
  expect(refNames).toEqual(["foo", "bar"]);
  expect(log).toBe("");
});

test("type inside fn with same name as fn", () => {
  // this will fail wgsl compilation, but as long as it doesn't hang the linker, we're ok
  const src = `
    fn foo() {
      let a:foo;
    }
    fn bar() {}
  `;
  const { refs, log } = traverseWithLog(src);
  expect(log).toBe("");
  expect(refs.length).toBe(2);
});

test("call inside fn with same name as fn", () => {
  const src = `
    fn foo() {
      foo();
    }
  `;
  const { refs, log } = traverseWithLog(src);
  expect(refs.length).toBe(1);
  expect(log).toBe("");
});

test("call cross reference", () => {
  const src = `
    fn foo() {
      bar();
    }

    fn bar() {
      foo();
    }
  `;
  const { refs, log } = traverseWithLog(src);
  const refNames = refs.map(r => (r as TextRef).elem.name);
  expect(refNames).toContain("foo");
  expect(refNames).toContain("bar");
  expect(refNames.length).toBe(2);
  expect(log).toBe("");
});

test("struct self reference", () => {
  const src = `
    struct A {
      a: A,
      b: B,
    }
    struct B {
      f: f32,
    }
  `;
  const { log } = traverseWithLog(src);
  expect(log).toBe("");
});

test("struct cross reference", () => {
  const src = `
    struct A {
      b: B,
    }
    struct B {
      a: A,
    }
  `;
  const { refs, log } = traverseWithLog(src);
  expect(log).toBe("");
  const refNames = refs.map(r => (r as any).elem.name);
  expect(refNames).toContain("A");
  expect(refNames).toContain("B");
  expect(refNames.length).toBe(2);
});

test("parse texture_storage_2d with texture format in type position", () => {
  const src = `var t: texture_storage_2d<rgba8unorm, write>;`;
  const { log } = traverseWithLog(src);
  expect(log).toBe("");
});

/** run traverseRefs with no filtering and return the refs and the error log output */
function traverseWithLog(
  src: string,
  ...modules: string[]
): { refs: FoundRef[]; log: string } {
  const { log, logged } = logCatch();
  const refs = _withBaseLogger(log, () => traverseTest(src, ...modules));

  return { refs, log: logged() };
}

/** run traverseRefs on the provided wgsl source strings
 * the first module is treated as the root
 */
function traverseTest(src: string, ...modules: string[]): FoundRef[] {
  const moduleFiles = Object.fromEntries(
    modules.map((m, i) => [`./file${i + 1}.wgsl`, m]),
  );
  const wgsl = { "./main.wgsl": src, ...moduleFiles };
  const registry = new ModuleRegistry({ wgsl });
  const refs: FoundRef[] = [];
  const parsed = registry.parsed();
  const mainModule = parsed.findTextModule("./main")!;
  const seen = new Set<string>();

  traverseRefs(mainModule, parsed, ref => {
    if (unseen(ref)) {
      refs.push(ref);
      return true;
    }
  });

  function unseen(ref: FoundRef): true | undefined {
    const fullName = refFullName(ref);
    if (!seen.has(fullName)) {
      seen.add(fullName);
      return true;
    }
  }

  return refs;
}
