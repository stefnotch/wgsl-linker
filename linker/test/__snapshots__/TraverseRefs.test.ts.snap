export const snapshot = {};

snapshot[`mismatched import export params 1`] = `
"mismatched import and export params  module: _root/main.wgsl
    #import foo(A, B) from ./file1   Ln 2
    ^
 module: _root/file1.wgsl
    #export(C)    Ln 2
    ^"
`;
